test/EResourceTest.cpp

```diff
@@ -0,0 +1,228 @@
+                }
+            }
+        }
+    }
+
+    GIVEN("The EResource locked using the lock function")
+    {
+        kf::EResource resource;
+        resource.lock();
+
+        THEN("The resource is acquired exclusively and shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == true);
+            REQUIRE(resource.isAcquiredShared() == 1);
+        }
+
+        WHEN("The resource is unlocked")
+        {
+            resource.unlock();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource is locked shared using the lock_shared function")
+    {
+        kf::EResource resource;
+        resource.lock_shared();
+
+        THEN("The resource is acquired shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == false);
+            REQUIRE(resource.isAcquiredShared() > 0);
+        }
+
+        WHEN("The resource is unlocked shared")
+        {
+            resource.unlock_shared();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource and 2 threads")
+    {
+        constexpr int kOneSecond = 10'000'000; // 1 second in 100-nanosecond intervals
+        struct Context
+        {
+            kf::EResource* resource;
+            bool wasAcquiredByFirst = false;
+            bool wasAcquiredBySecond = false;
+        };
+
+        WHEN("Multiple threads attempt to acquire exclusive the resource")
+        {
+            kf::EResource resource;
+            kf::Thread thread1;
+            kf::Thread thread2;
+
+            Context context{ &resource };
+
+            thread1.start([](void* context) {
+                auto res = static_cast<Context*>(context);
+                res->wasAcquiredByFirst = res->resource->acquireExclusive();
+                res->resource->release();
+                },
+                &context);
+
+            thread2.start([](void* context) {
+                auto res = static_cast<Context*>(context);
+                res->wasAcquiredBySecond = res->resource->acquireExclusive();
+                res->resource->release();
+                },
+                &context);
+
+            thread1.join();
+            thread2.join();
+
+            THEN("Both threads can acquire and release the resource without deadlock")
+            {
+                REQUIRE(context.wasAcquiredByFirst);
+                REQUIRE(context.wasAcquiredBySecond);
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+
+        WHEN("Multiple threads attempt to acquire shared the resource")
```

Comment:
There is a better way to check that shared access was successfull without waiting. Create a hundred of threads that read some resource and register this event, then join them all and then check that every consumer got what it wanted